WordPress/WordPressTest/SiteCreation/SiteCreationNameTracksEventTests.swift

```diff
@@ -21,6 +21,47 @@ class SiteCreationNameTracksEventTests: XCTestCase {
         return siteIntentViewController
     }
 
+    class MockSiteNameStep: WizardStep {
+        weak var delegate: WizardDelegate?
+        private let creator: SiteCreator
+
+        var expectedProperty = ""
+
+        var content: UIViewController {
+            SiteNameViewController(siteNameViewFactory: makeSiteNameView) { [weak self] in
+                SiteCreationAnalyticsHelper.trackSiteNameSkipped()
+                self?.didSet(siteName: nil)
+            }
+        }
+
+        init(creator: SiteCreator) {
+            self.creator = creator
+        }
+
+        private func didSet(siteName: String?) {
+            if let siteName = siteName {
+                SiteCreationAnalyticsHelper.trackSiteNameEntered(siteName)
```

Comment:
I am not sure I follow: `testSiteNameTracksEventFiresWhenEntered` uses `mockSiteNameViewControllerMaker` which, in turn, instantiates `MockSiteNameStep`, and, if I remove that line from `MockSiteNameStep` the test fails.
If I remove it from `SiteNameStep` the test will indeed pass, but because it's not using it, but it's using the mock. Am I missing something?

Reply:
I see your point. the thing is that, ideally, we would have wanted to either inherit the mock from the original `SiteNameStep` (which currently is marked as `final`), or improve the injection so that whatever mock we inject into it would use that same call. The first one looks easier, but would break the current architecture of having all the "steps" final (although I am not sure if removing this condition would do any harm in an app?). the other one might be trickier, and it would still not be consistent with the overall architecture? Which one do you think it's better? Or do you have other approaches to suggest?

Reply:
The cleanest approach I landed on was to basically turn `SiteNameStep` into a view model. I followed your closure pattern on `SiteNameViewController` and it got the job done, but as it grew it screamed to me protocol + delegate pattern.

My vote would be to make `SiteNameViewController` a delegate so we knew it would conform to a protocol for `didView`, `didSkip`, and `didCancel`. (One hitch is that the view itself is handling `didSet`, but I would consider having that flow through the VC as well.)

Then in our tests we'd just create a mock type that conformed and cut out view code altogether. (One could even create separate tests that the view called those delegate methods.)

```
        return SiteNameViewController(siteNameViewFactory: makeSiteNameView) {
            weakSelf?.didView()
        } onSkip: {
            weakSelf?.didSkip()
        } onCancel: {
            weakSelf?.didCancel()
        }
```

One other benefit to this approach I noticed was that it was more precise on firing one event only rather than comparing against the latest event. This is minor, but feels a little more controlled.

[Here's the idea in a branch.](https://github.com/wordpress-mobile/WordPress-iOS/compare/task/site-name-tracks-idea?expand=1)

I'm also open to us going to the drawing board to figure out a way to support testing for all steps. 👍

Reply:
> I'm also open to us going to the drawing board to figure out a way to support testing for all steps.


Exactly so: I feel that we are going to a route of redesign/re-architecture, and it should probably be done outside the scope of this PR.

Maybe we could merge this PR (maybe with just the change of not making `SiteNameStep` final and having a more proper mock), and then we discuss how to improve testability for the whole site creation, what do you think?

Reply:
Yep, sounds like a way to move forward. 👍

Reply:
And actually just to add one more note, it could be even easier than a delegate pattern if the VC took in a `SiteNameStep`.

Reply:
I pushed a change for this. Looking again into this, I realized that we can use the same technique of getting a reference to the UI object that passes the `siteName` property: this was, we don't need an unrelated mock, and we don't even need to make any changes to `SiteNameStep`!